README.md
# todo-or-die

`todo-or-die` provides procedural macros that act as checked reminders.

[![Build status](https://github.com/davidpdrsn/todo-or-die/actions/workflows/CI.yml/badge.svg?branch=main)](https://github.com/davidpdrsn/todo-or-die/actions/workflows/CI.yml)
[![Crates.io](https://img.shields.io/crates/v/todo-or-die)](https://crates.io/crates/todo-or-die)
[![Documentation](https://docs.rs/todo-or-die/badge.svg)](https://docs.rs/todo-or-die)

# Examples

```rust
// trigger a compile error if we're past a certain date
todo_or_die::after_date!(3000, 1, 1); // its the year 3000!

// or a GitHub issue has closed
todo_or_die::issue_closed!("rust-lang", "rust", 44265); // GATs are here!

// or the latest version of a crate matches some expression
todo_or_die::crates_io!("serde", ">1.0.9000"); // its over 9000!
```

More information about this crate can be found in the [crate documentation][docs].

The name was shamelessly stolen from the ruby gem [`todo_or_die`][ruby].

## License

This project is licensed under the [MIT license](LICENSE).

[ruby]: https://rubygems.org/gems/todo_or_die
[docs]: https://docs.rs/todo-or-die


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# Unreleased

- None.

# 0.1.2 (17. September, 2021)

- Timeout HTTP requests after 1 second.

# 0.1.1 (06. September, 2021)

- Add `rust_version` for checking the currently active version of rust.
- Cache HTTP requests. Requests will be cached for 1 hour by default. This can
  be customized by setting `TODO_OR_DIE_HTTP_CACHE_TTL_SECONDS`.

# 0.1.0 (05. September, 2021)

- Initial release.


src/http.rs
use anyhow::{Context as _, Result};
use chrono::prelude::*;
use hyper::{
    body::Bytes,
    client::{connect::dns::GaiResolver, HttpConnector},
    header::HeaderValue,
    header::USER_AGENT,
    Body, Client, Request, Response,
};
use hyper_rustls::HttpsConnector;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::{
    collections::{hash_map::DefaultHasher, HashMap},
    hash::{Hash, Hasher},
    path::PathBuf,
};
use tokio::runtime::Runtime;

pub(crate) fn request<T>(
    // the request body isn't used in the cache key, so require it to be `()` so
    // we can guarantee that its empty
    request: Request<()>,
) -> Result<T>
where
    T: serde::de::DeserializeOwned,
{
    RUNTIME.block_on(async move {
        if should_clear_cache() {
            clear_cache().ok();
        }

        let mut request = request.map(|_| Body::empty());

        request
            .headers_mut()
            .insert(USER_AGENT, HeaderValue::from_static("todo-or-die"));

        let hash = hash_request(&request);

        let response = if let Some(cached_response) =
            cached_response(&hash).context("Failed to read cached response")?
        {
            cached_response
        } else {
            execute_request_and_cache_response(request, &hash).await?
        };

        if !response.status().is_success() {
            let body = String::from_utf8_lossy(response.body());
            anyhow::bail!(
                "Received non-success response. status={}, body={:?}",
                response.status(),
                body
            );
        }

        let value =
            serde_json::from_slice::<T>(&*response.body()).context("Failed to parse response")?;
        Ok(value)
    })
}

async fn execute_request_and_cache_response(
    request: Request<Body>,
    hash: &RequestHash,
) -> Result<Response<Bytes>> {
    let response = tokio::time::timeout(
        std::time::Duration::from_secs(1),
        http_client().request(request),
    )
    .await
    .context("HTTP request timed out")?
    .context("HTTP request to failed")?;

    let (parts, body) = response.into_parts();
    let body = hyper::body::to_bytes(body)
        .await
        .context("Failed to read response")?;
    let response = Response::from_parts(parts, body);

    if caching_enabled() {
        cache_response(hash, &response).context("Failed to cache response")?;
    }

    Ok(response)
}

static RUNTIME: Lazy<Runtime> = Lazy::new(|| {
    tokio::runtime::Builder::new_current_thread()
        .enable_all()
        .build()
        .expect("failed to build tokio runtime")
});

type HyperTlsClient = Client<HttpsConnector<HttpConnector<GaiResolver>>, Body>;

fn http_client() -> &'static HyperTlsClient {
    static CLIENT: Lazy<HyperTlsClient> = Lazy::new(|| {
        let mut tls = rustls::ClientConfig::new();
        tls.set_protocols(&["h2".into(), "http/1.1".into()]);
        tls.root_store
            .add_server_trust_anchors(&webpki_roots::TLS_SERVER_ROOTS);

        let mut http = hyper::client::HttpConnector::new();
        http.enforce_http(false);

        hyper::Client::builder().build::<_, Body>(hyper_rustls::HttpsConnector::from((http, tls)))
    });

    &*CLIENT
}

#[derive(Clone, Debug, PartialEq, Eq)]
struct RequestHash(String);

fn hash_request(request: &Request<Body>) -> RequestHash {
    let mut hasher = DefaultHasher::new();
    format!("{:?}", request).hash(&mut hasher);
    let hash = hasher.finish();
    RequestHash(hash.to_string())
}

fn cached_response(hash: &RequestHash) -> Result<Option<Response<Bytes>>> {
    if !caching_enabled() {
        return Ok(None);
    }

    let path = cache_dir_path_for_this_version()?.join(&hash.0);

    let data = match std::fs::read(&path) {
        Ok(file) => file,
        Err(err) if err.kind() == std::io::ErrorKind::NotFound => {
            return Ok(None);
        }
        Err(err) => return Err(err.into()),
    };

    if let Some(response) = deserialize_response(data)? {
        Ok(Some(response))
    } else {
        std::fs::remove_file(&path)?;
        Ok(None)
    }
}

fn cache_response(hash: &RequestHash, response: &Response<Bytes>) -> Result<()> {
    let path = cache_dir_path_for_this_version()?.join(&hash.0);
    let bytes = serialize_response(response)?;
    std::fs::write(path, bytes)?;
    Ok(())
}

fn clear_cache() -> Result<()> {
    let path = top_level_cache_dir()?;
    std::fs::remove_dir_all(path)?;
    Ok(())
}

fn serialize_response(response: &Response<Bytes>) -> Result<Vec<u8>> {
    let headers = response
        .headers()
        .iter()
        .map(|(key, value)| (key.as_str().to_string(), value.as_bytes().to_vec()))
        .collect();

    let response = SerializedResponse {
        status: response.status().as_u16(),
        headers,
        body: response.body().to_vec(),
        expires_at: Local::now() + cache_ttl(),
    };

    Ok(serde_json::to_vec(&response)?)
}

fn cache_ttl() -> chrono::Duration {
    (|| {
        let var = std::env::var("TODO_OR_DIE_HTTP_CACHE_TTL_SECONDS")?;
        let sec = var.parse()?;
        Ok::<_, anyhow::Error>(chrono::Duration::seconds(sec))
    })()
    .unwrap_or_else(|_| chrono::Duration::hours(1))
}

fn deserialize_response(data: Vec<u8>) -> Result<Option<Response<Bytes>>> {
    let response = serde_json::from_slice::<SerializedResponse>(&data)
        .context("Failed to deserialize cached HTTP response")?;

    let expires_at = response.expires_at.timestamp();
    let now = Local::now().timestamp();
    if now > expires_at {
        return Ok(None);
    }

    let status = hyper::StatusCode::from_u16(response.status)?;

    let headers = response
        .headers
        .iter()
        .map(|(key, value)| {
            let key = hyper::header::HeaderName::from_bytes(key.as_bytes())?;
            let value = HeaderValue::from_bytes(value)?;
            Ok::<_, anyhow::Error>((key, value))
        })
        .collect::<Result<hyper::HeaderMap>>()
        .context("Failed to build header map")?;

    let body = Bytes::copy_from_slice(&response.body);

    let mut out = Response::new(body);
    *out.status_mut() = status;
    *out.headers_mut() = headers;
    Ok(Some(out))
}

#[derive(Serialize, Deserialize)]
struct SerializedResponse {
    status: u16,
    headers: HashMap<String, Vec<u8>>,
    body: Vec<u8>,
    expires_at: DateTime<Local>,
}

fn top_level_cache_dir() -> Result<PathBuf> {
    let path = std::env::temp_dir().join("todo_or_die_cache");
    std::fs::create_dir_all(&path).context("Failed to create dir to store HTTP caches")?;
    Ok(path)
}

fn cache_dir_path_for_this_version() -> Result<PathBuf> {
    let todo_or_die_version = env!("CARGO_PKG_VERSION");
    let path = top_level_cache_dir()?.join(todo_or_die_version);
    std::fs::create_dir_all(&path).context("Failed to create dir to store HTTP caches")?;
    Ok(path)
}

fn caching_enabled() -> bool {
    !should_clear_cache() && std::env::var("TODO_OR_DIE_DISABLE_HTTP_CACHE").is_err()
}

fn should_clear_cache() -> bool {
    std::env::var("TODO_OR_DIE_CLEAR_HTTP_CACHE").is_ok()
}


src/rust.rs
use anyhow::{Context as _, Result};
use semver::VersionReq;
use syn::parse::Parse;
use version_check::Version;

pub(crate) fn rust_version(input: Input) -> Result<Option<String>> {
    let current_version = Version::read()
        .context("Unable to get current rust version")?
        .to_string()
        .parse::<semver::Version>()
        .context("Couldn't parse rust version")?;

    if input.version_req.matches(&current_version) {
        Ok(Some(format!(
            "Your active version of rust is {}. Time to act on this!",
            current_version
        )))
    } else {
        Ok(None)
    }
}

pub(crate) struct Input {
    version_req: VersionReq,
}

impl Parse for Input {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let lit = input.parse::<syn::LitStr>()?;
        let version_req = lit
            .value()
            .parse()
            .map_err(|err| syn::Error::new(lit.span(), err))?;

        input.parse::<syn::token::Comma>().ok();

        Ok(Self { version_req })
    }
}

/// ```compile_fail
/// todo_or_die::rust_version!(">1.50");
/// ```
///
/// ```
/// todo_or_die::rust_version!("=2.0.0");
/// ```
#[allow(dead_code)]
fn tests() {}


src/time.rs
use anyhow::Result;
use chrono::prelude::*;
use syn::parse::Parse;

pub(crate) fn after_date(input: Input) -> Result<Option<String>> {
    let input = NaiveDate::from_ymd(input.year, input.month, input.day);
    let today = Local::today().naive_local();

    if input <= today {
        Ok(Some(format!(
            "{} is now in the past. Time to act on this!",
            input
        )))
    } else {
        Ok(None)
    }
}

pub(crate) struct Input {
    year: i32,
    month: u32,
    day: u32,
}

impl Parse for Input {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let year = input.parse::<syn::LitInt>()?.base10_parse()?;
        input.parse::<syn::token::Comma>()?;

        let month = input.parse::<syn::LitInt>()?.base10_parse()?;
        input.parse::<syn::token::Comma>()?;

        let day = input.parse::<syn::LitInt>()?.base10_parse()?;

        input.parse::<syn::token::Comma>().ok();

        Ok(Self { year, month, day })
    }
}

/// ```compile_fail
/// todo_or_die::after_date!(1990, 01, 01);
/// ```
///
/// ```
/// todo_or_die::after_date!(3000, 01, 01);
/// ```
#[allow(dead_code)]
fn tests() {}


src/krate.rs
use crate::http::request;
use anyhow::{Context as _, Result};
use hyper::Request;
use semver::{Version, VersionReq};
use serde::Deserialize;
use syn::parse::Parse;

pub(crate) fn crates_io(input: Input) -> Result<Option<String>> {
    #[derive(Debug, Deserialize)]
    struct Response {
        versions: Vec<CrateVersion>,
    }

    #[derive(Debug, Deserialize)]
    struct CrateVersion {
        num: String,
    }

    let data = request::<Response>(
        Request::builder()
            .uri(format!("https://crates.io/api/v1/crates/{}", input.krate))
            .body(())
            .unwrap(),
    )?;

    let latest_version = data
        .versions
        .first()
        .context("No versions found for crate")?
        .num
        .parse::<Version>()?;

    if input.version_req.matches(&latest_version) {
        Ok(Some(format!(
            "Latest version of {} is {}. Time to act on this!",
            input.krate, latest_version
        )))
    } else {
        Ok(None)
    }
}

pub(crate) struct Input {
    krate: String,
    version_req: VersionReq,
}

impl Parse for Input {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let krate = input.parse::<syn::LitStr>()?.value();

        input.parse::<syn::token::Comma>()?;

        let lit = input.parse::<syn::LitStr>()?;
        let version_req = lit
            .value()
            .parse()
            .map_err(|err| syn::Error::new(lit.span(), err))?;

        input.parse::<syn::token::Comma>().ok();

        Ok(Self { krate, version_req })
    }
}

/// ```compile_fail
/// todo_or_die::crates_io!("tokio", ">=1.0");
/// ```
///
/// ```
/// todo_or_die::crates_io!("tokio", ">=10.0");
/// ```
#[allow(dead_code)]
fn tests() {}


src/github.rs
use crate::http::request;
use anyhow::{Context as _, Result};
use hyper::{
    header::HeaderValue,
    header::{ACCEPT, AUTHORIZATION},
    Request,
};
use serde::Deserialize;
use syn::parse::Parse;

pub(crate) fn issue_closed(input: OrgRepoIssue) -> Result<Option<String>> {
    #[derive(Deserialize, Debug)]
    struct Issue {
        closed_at: Option<String>,
    }

    let OrgRepoIssue {
        org,
        repo,
        issue: issue_number,
    } = input;

    let issue = request::<Issue>(github_request(
        Request::builder()
            .uri(format!(
                "https://api.github.com/repos/{}/{}/issues/{}",
                org, repo, issue_number
            ))
            .body(())
            .unwrap(),
    )?)?;

    if issue.closed_at.is_some() {
        Ok(Some(format!(
            "{}/{}#{} is closed. Time to act on this!",
            org, repo, issue_number
        )))
    } else {
        Ok(None)
    }
}

#[derive(Deserialize, Debug)]
struct PullRequest {
    state: String,
}

pub(crate) fn pr_closed(input: OrgRepoIssue) -> Result<Option<String>> {
    let OrgRepoIssue {
        org,
        repo,
        issue: issue_number,
    } = input;

    let pr = request::<PullRequest>(github_request(
        Request::builder()
            .uri(format!(
                "https://api.github.com/repos/{}/{}/pulls/{}",
                org, repo, issue_number
            ))
            .body(())
            .unwrap(),
    )?)?;

    if pr.state == "closed" {
        Ok(Some(format!(
            "{}/{}#{} is closed. Time to act on this!",
            org, repo, issue_number
        )))
    } else {
        Ok(None)
    }
}

fn github_request<B>(mut request: Request<B>) -> Result<Request<B>> {
    request.headers_mut().insert(
        ACCEPT,
        HeaderValue::from_static("application/vnd.github.v3+json"),
    );

    if let Some(auth_token) = auth_token() {
        request.headers_mut().insert(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {}", auth_token))
                .context("GitHub auth token contained invalid header value")?,
        );
    }

    Ok(request)
}

fn auth_token() -> Option<String> {
    std::env::var("TODO_OR_DIE_GITHUB_TOKEN")
        .ok()
        .or_else(|| std::env::var("GITHUB_TOKEN").ok())
}

pub(crate) struct OrgRepoIssue {
    org: String,
    repo: String,
    issue: u64,
}

impl Parse for OrgRepoIssue {
    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
        let org = input.parse::<syn::LitStr>()?.value();
        input.parse::<syn::token::Comma>()?;

        let repo = input.parse::<syn::LitStr>()?.value();
        input.parse::<syn::token::Comma>()?;

        let issue = input.parse::<syn::LitInt>()?.base10_parse()?;

        input.parse::<syn::token::Comma>().ok();

        Ok(Self { org, repo, issue })
    }
}

/// # `issue_closed`
///
/// closed issue
/// ```compile_fail
/// todo_or_die::issue_closed!("tokio-rs", "axum", 1);
/// ```
///
/// open issue
/// ```
/// // the oldest open rust-lang issue. Probably wont be close anytime soon :shrug:
/// todo_or_die::issue_closed!("rust-lang", "rust", 1563);
/// ```
///
/// # `pr_closed`
///
/// closed pr
/// ```compile_fail
/// todo_or_die::pr_closed!("tokio-rs", "axum", 266);
/// ```
///
/// merged pr
/// ```compile_fail
/// todo_or_die::pr_closed!("tokio-rs", "axum", 294);
/// ```
///
/// open pr
/// ```
/// todo_or_die::pr_closed!("davidpdrsn", "keep", 1);
/// ```
#[allow(dead_code)]
fn tests() {}


src/lib.rs
//! TODOs you cannot forget!
//!
//! `todo-or-die` provides procedural macros that act as checked reminders.
//!
//! The name was shamelessly stolen from the ruby gem [`todo_or_die`][ruby].
//!
//! # Examples
//!
//! ```
//! // trigger a compile error if we're past a certain date
//! todo_or_die::after_date!(3000, 1, 1); // its the year 3000!
//!
//! // or a GitHub issue has closed
//! todo_or_die::issue_closed!("rust-lang", "rust", 44265); // GATs are here!
//!
//! // or the latest version of a crate matches some expression
//! todo_or_die::crates_io!("serde", ">1.0.9000"); // its over 9000!
//! ```
//!
//! # Feature flags
//!
//! The following optional features are available:
//!
//! - `crate`: Enables checking versions of crates.
//! - `github`: Enables checking if issues or pull requests are closed.
//! - `rust`: Enables checking the current rust version.
//! - `time`: Enables checking things to do with time.
//!
//! Note that _none_ of the features are enabled by default.
//!
//! # Skipping checks
//!
//! If the environment variable `TODO_OR_DIE_SKIP` is set all macros will do
//! nothing and immediately succeed. This can for example be used to skip checks
//! locally and only perform them on CI.
//!
//! # Caching HTTP requests
//!
//! By default HTTP requests will be cached. The behavior can be customized with
//! these environment variables:
//! - `TODO_OR_DIE_HTTP_CACHE_TTL_SECONDS`: How long cached responses will be
//! used. The default is 1 hour.
//! - `TODO_OR_DIE_DISABLE_HTTP_CACHE`: Disables caching if its set.
//! - `TODO_OR_DIE_CLEAR_HTTP_CACHE`: Clears the cache if its set.
//!
//! The cache is stored at `std::env::temp_dir().join("todo_or_die_cache")`.
//!
//! # You can still compile offline
//!
//! If you're offline or GitHub is down you can still build. If the macros hit
//! some kind of error a warning will be printed but they wont trigger a compile
//! error.
//!
//! [ruby]: https://rubygems.org/gems/todo_or_die

#![warn(
    clippy::all,
    clippy::dbg_macro,
    clippy::todo,
    clippy::empty_enum,
    clippy::enum_glob_use,
    clippy::mem_forget,
    clippy::unused_self,
    clippy::filter_map_next,
    clippy::needless_continue,
    clippy::needless_borrow,
    clippy::match_wildcard_for_single_variants,
    clippy::if_let_mutex,
    clippy::mismatched_target_os,
    clippy::await_holding_lock,
    clippy::match_on_vec_items,
    clippy::imprecise_flops,
    clippy::suboptimal_flops,
    clippy::lossy_float_literal,
    clippy::rest_pat_in_fully_bound_structs,
    clippy::fn_params_excessive_bools,
    clippy::exit,
    clippy::inefficient_to_string,
    clippy::linkedlist,
    clippy::macro_use_imports,
    clippy::option_option,
    clippy::verbose_file_reads,
    clippy::unnested_or_patterns,
    rust_2018_idioms,
    future_incompatible,
    nonstandard_style,
    missing_debug_implementations,
    missing_docs
)]
#![deny(unreachable_pub, private_in_public)]
#![allow(elided_lifetimes_in_paths, clippy::type_complexity)]
#![forbid(unsafe_code)]
#![cfg_attr(docsrs, feature(doc_cfg))]
#![cfg_attr(test, allow(clippy::float_cmp))]

#[cfg(feature = "__internal_http")]
mod http;

#[cfg(feature = "github")]
mod github;

#[cfg(feature = "time")]
mod time;

#[cfg(feature = "crate")]
mod krate;

#[cfg(feature = "rust")]
mod rust;

/// Trigger a compile error if an issue has been closed.
///
/// Note that this will make network requests during compile which may make your builds flaky at
/// times.
///
/// Requires the `github` feature to be enabled.
///
/// # Example
///
/// ```compile_fail
/// todo_or_die::issue_closed!("tokio-rs", "axum", 1);
/// ```
///
/// # Authentication
///
/// `issue_closed` will first look for the environment variable `TODO_OR_DIE_GITHUB_TOKEN` and then
/// `GITHUB_TOKEN`, if either are found its value will be used as the auth token when making
/// requests to the GitHub API. This allows you to access private repos and get more generous
/// rate limits.
#[cfg(feature = "github")]
#[proc_macro]
pub fn issue_closed(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    perform_check(input, github::issue_closed)
}

/// Trigger a compile error if a pull request has been closed or merged.
///
/// Note that this will make network requests during compile which may make your builds flaky at
/// times.
///
/// Requires the `github` feature to be enabled.
///
/// # Example
///
/// ```compile_fail
/// todo_or_die::pr_closed!("tokio-rs/axum#266");
/// // todo_or_die::pr_closed!("tokio-rs", "axum", 266);
/// ```
///
/// # Authentication
///
/// `pr_closed` will first look for the environment variable `TODO_OR_DIE_GITHUB_TOKEN` and then
/// `GITHUB_TOKEN`, if either are found its value will be used as the auth token when making
/// requests to the GitHub API. This allows you to access private repos and get more generous
/// rate limits.
#[cfg(feature = "github")]
#[proc_macro]
pub fn pr_closed(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    perform_check(input, github::pr_closed)
}

/// Trigger a compile error if today is after the given date
///
/// Requires the `time` feature to be enabled.
///
/// # Example
///
/// ```compile_fail
/// todo_or_die::after_date!(1994, 10, 22);
/// ```
#[cfg(feature = "time")]
#[proc_macro]
pub fn after_date(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    perform_check(input, time::after_date)
}

/// Trigger a compile error if the latest version of a crate hosted on crates.io matches some
/// expression.
///
/// Note that this will make network requests during compile which may make your builds flaky at
/// times.
///
/// Requires the `crate` feature to be enabled.
///
/// # Example
///
/// ```compile_fail
/// todo_or_die::crates_io!("tokio", ">=1.0");
/// ```
///
/// Any version requirement supported by [`semver::VersionReq::parse`] is supported.
///
/// [`semver::VersionReq::parse`]: https://docs.rs/semver/latest/semver/struct.VersionReq.html#method.parse
#[cfg(feature = "crate")]
#[proc_macro]
pub fn crates_io(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    perform_check(input, krate::crates_io)
}

/// Trigger a compile error if the currently used version of rust used matches some expression.
///
/// Note that release channels (like `nightly` or `beta`) are ignored.
///
/// Requires the `rust` feature to be enabled.
///
/// # Example
///
/// ```compile_fail
/// todo_or_die::rust_version!(">1.50");
/// ```
///
/// Any version requirement supported by [`semver::VersionReq::parse`] is supported.
///
/// [`semver::VersionReq::parse`]: https://docs.rs/semver/latest/semver/struct.VersionReq.html#method.parse
#[cfg(feature = "rust")]
#[proc_macro]
pub fn rust_version(input: proc_macro::TokenStream) -> proc_macro::TokenStream {
    perform_check(input, rust::rust_version)
}

#[allow(dead_code)]
fn perform_check<F, T>(input: proc_macro::TokenStream, f: F) -> proc_macro::TokenStream
where
    F: FnOnce(T) -> anyhow::Result<Option<String>>,
    T: syn::parse::Parse,
{
    if std::env::var("TODO_OR_DIE_SKIP").is_ok() {
        return Default::default();
    }

    let input = match syn::parse::<T>(input) {
        Ok(value) => value,
        Err(err) => {
            let err = err.to_string();
            return quote::quote! {
                ::std::compile_error!(#err);
            }
            .into();
        }
    };

    match f(input) {
        Ok(None) => {}
        Ok(Some(msg)) => {
            return quote::quote! {
                ::std::compile_error!(#msg);
            }
            .into();
        }
        Err(err) => {
            eprintln!("something went wrong\n\n{:?}", err);
        }
    }

    Default::default()
}
